Content.Server/Database/PostgresNotificationManager.cs

```diff
@@ -0,0 +1,162 @@
+/// Players a ban_notification got received for get banned, except when the current server id and the one in the notification payload match.
+/// </summary>
+public sealed class PostgresNotificationManager : IDisposable
+{
+    private const string BanNotificationChannel = "ban_notification";
+    private const string PostgresDbEngine = "postgres";
+
+    [Dependency] private readonly ITaskManager _taskManager = default!;
+    [Dependency] private readonly IServerNetManager _netManager = default!;
+    [Dependency] private readonly ISharedPlayerManager _playerManager = default!;
+    [Dependency] private readonly ILogManager _logManager = default!;
+    [Dependency] private readonly ILocalizationManager _loc = default!;
+    [Dependency] private readonly IConfigurationManager _cfg = default!;
+    [Dependency] private readonly ServerDbEntryManager _entryManager = default!;
+
+#if EXCEPTION_TOLERANCE
+    [Dependency] private readonly IRuntimeLog _runtimeLog = default!;
+#endif
+
+    private ISawmill? _logger;
+    private NpgsqlConnection? _connection;
+
+    private readonly CancellationTokenSource _tokenSource = new();
+
+    /// <summary>
+    /// Sets up the database connection and the notification handler
+    /// </summary>
+    public void Init()
+    {
+        if (!_cfg.GetCVar(CCVars.DatabaseEngine).Equals(PostgresDbEngine, StringComparison.CurrentCultureIgnoreCase))
+            return;
+
+        _logger = _logManager.GetSawmill("db.listener");
+
+        var host = _cfg.GetCVar(CCVars.DatabasePgHost);
+        var port = _cfg.GetCVar(CCVars.DatabasePgPort);
+        var db = _cfg.GetCVar(CCVars.DatabasePgDatabase);
+        var user = _cfg.GetCVar(CCVars.DatabasePgUsername);
+        var pass = _cfg.GetCVar(CCVars.DatabasePgPassword);
+
+        var builder = new NpgsqlConnectionStringBuilder
+        {
+            Host = host,
+            Port = port,
+            Database = db,
+            Username = user,
+            Password = pass
+        };
+
+        _connection = new NpgsqlConnection(builder.ConnectionString);
+        _connection.Notification += OnNotification;
+
+        var cancellationToken = _tokenSource.Token;
+        Task.Run(() => NotificationListener(cancellationToken), cancellationToken);
+    }
+
+    /// <summary>
+    /// Listens to the notification channel with basic error handling and reopens the connection if it got closed
+    /// </summary>
+    private async Task NotificationListener(CancellationToken cancellationToken)
+    {
+        if (_connection == null)
+            return;
+
+        while (!cancellationToken.IsCancellationRequested)
+        {
+            if (_connection.State == ConnectionState.Broken)
+                await _connection.CloseAsync();
+
+            if (_connection.State == ConnectionState.Closed)
+            {
+                await _connection.OpenAsync(cancellationToken);
```

Comment:
The npgsql documentation mentions something about keepalives, is that necessary?

Reply:
This might not be necessary. The documentation says that you should try to do it without first.

> Note that you should only turn this feature on if you need it. Unless you know you'll have long-lived idle connections, and that your backend (or network equipment) will interfere with these connections, it's better to leave this off.

I don't know if anything would interfere with long lived connections so we probably don't need it.

Reply:
Right, reading that it's probably not necessary.